### PR TITLE
feat(tasks): add visual indicator for unsynced local changes

### PR DESCRIPTION
**Pull request by ShivaGupta-14**
### Description

This PR implements **visual indicators for unsynced task modifications**. When a task is edited, completed, or deleted, the UI immediately marks it as **"Unsynced"** until the backend successfully synchronizes all pending changes.

- Closes: #143

### Key Changes
- **State Management:** Implements a local state `unsyncedTaskUuids` to track tasks modified on the frontend but not yet synced to the backend.
- **Visual Indicators:**
  - Adds a **red "Unsynced" badge** to task rows for immediate feedback when a task is edited, completed, or deleted.
  - Displays the **count of pending unsynced items** on the Sync button via a notification-style badge.
- **Optimistic UI Updates:** Immediately updates UI for task status changes (completed/deleted) to enhance perceived performance.
- **Sync Logic:** Automatically clears all unsynced indicators once backend synchronization succeeds.

- **Testing**
  - Includes unit tests verifying:
    - Unsynced badge visibility  
    - Sync button counter accuracy  
    - Clearing state on successful sync
    - Used `test.each` for repetitive date editing tests, reducing code duplication.

### Checklist

- [x] Ran `npx prettier --write .` (for formatting)
- [ ] Ran `gofmt -w .` (for Go backend)
- [x] Ran `npm test` (for JS/TS testing)
- [x] Added unit tests, if applicable
- [x] Verified all tests pass
- [ ] Updated documentation, if needed

### Additional Notes

Video: [Link to see changes](https://drive.google.com/file/d/1lJoPKuyjC2lyuQ0KqcBKWKXhxW_D0yjL/view?usp=sharing)

### CODE REVIEW

File: frontend/src/components/HomeComponents/Tasks/Tasks.tsx [hunk 1]
```diff
@@ -339,6 +342,39 @@ export const Tasks = (
   ) {
     if (handleDate(newTask.due)) {
       try {
+        const tempTask: Task = {
```
Comment (its-me-abhishek) [inline frontend/src/components/HomeComponents/Tasks/Tasks.tsx]: why'd we need a tempTask?

File: frontend/src/components/HomeComponents/Tasks/__tests__/Tasks.test.tsx [hunk 2]
```diff
@@ -421,4 +438,134 @@ describe('Tasks Component', () => {
...
+    await waitFor(() => {
+      const yesButtons = screen.getAllByText('Yes');
+      if (yesButtons.length > 0) fireEvent.click(yesButtons[0]);
+    });
+
+    await waitFor(() => {
+      expect(screen.getByText('Unsynced')).toBeInTheDocument();
+    });
+  });
+
```
Comment (its-me-abhishek) [inline frontend/src/components/HomeComponents/Tasks/__tests__/Tasks.test.tsx]: misses edge cases where Unsynced badge shall be shown in case any task field is edited

**Review by its-me-abhishek — CHANGES_REQUESTED** [review]
hey, this does not really go along the lines that is intended for this feature. The feature shall simply just be a state of set that handles a set of UUIDs of tasks that have been recently modified, and hence display them using the set calls upon rendering.

**Comment by ShivaGupta-14** [comment]
@its-me-abhishek, Thanks for explaining! I've made the necessary changes, the "Unsynced" badge will now only appear when an existing task is edited, not when a new task is added.

Quick question: Should I update the toast from "Task added successfully!" to "Task added! Sync to see it (or update)." to make it clearer? Or keep it as is?

PR is ready for review.

File: frontend/src/components/HomeComponents/Tasks/__tests__/Tasks.test.tsx [hunk 3]
```diff
@@ -493,4 +510,397 @@ describe('Tasks Component', () => {
 
     expect(newProjectInput).toHaveValue('My Fresh Project');
   });
+
+  // NEW TESTS FOR UNSYNCED FEATURE
```
Comment (its-me-abhishek) [inline frontend/src/components/HomeComponents/Tasks/__tests__/Tasks.test.tsx]: this comment shall be removed. this can be a comment on self-review but not in the codebase directly...

File: frontend/src/components/HomeComponents/Tasks/__tests__/Tasks.test.tsx [hunk 4]
```diff
@@ -493,4 +510,397 @@ describe('Tasks Component', () => {
...
+    ]);
+
+    const syncButtons = screen.getAllByText('Sync');
+    fireEvent.click(syncButtons[0]);
+
+    await waitFor(
+      () => {
+        expect(screen.queryByText('Unsynced')).not.toBeInTheDocument();
+      },
+      { timeout: 3000 }
```
Comment (its-me-abhishek) [inline frontend/src/components/HomeComponents/Tasks/__tests__/Tasks.test.tsx]: is this constant arbitrary?

**Review by its-me-abhishek — CHANGES_REQUESTED** [review]

File: frontend/src/components/HomeComponents/Tasks/TaskDialog.tsx [hunk 5]
```diff
@@ -108,7 +109,7 @@ export const TaskDialog = ({
                   : 'dark:text-white text-black'
               }`}
             >
-              {task.id}
+              {task.id < 0 ? '-' : task.id}
```
[the remark this answers is not in the document]
Comment (ShivaGupta-14) [reply frontend/src/components/HomeComponents/Tasks/TaskDialog.tsx]: fixed! I had removed the negative case earlier but missed updating this part during the recent changes

File: frontend/src/components/HomeComponents/Tasks/__tests__/Tasks.test.tsx [hunk 6]
```diff
@@ -90,6 +90,16 @@ jest.mock('../hooks', () => ({
               tags: i % 3 === 0 ? ['tag1'] : ['tag2'],
               uuid: `uuid-${i + 1}`,
               due: i === 0 ? '20200101T120000Z' : undefined,
+              entry: '',
```
[the remark this answers is not in the document]
Comment (ShivaGupta-14) [reply frontend/src/components/HomeComponents/Tasks/__tests__/Tasks.test.tsx]: I added these for type completeness to match the Task structure, but verified tests pass without them. Removed now.

File: frontend/src/components/HomeComponents/Tasks/TaskDialog.tsx
Comment (ShivaGupta-14) [inline frontend/src/components/HomeComponents/Tasks/TaskDialog.tsx]: shows animated unsynced badge when task has local changes pending sync. Added aria-labels for accessibility

File: frontend/src/components/HomeComponents/Tasks/Tasks.tsx
Comment (ShivaGupta-14) [inline frontend/src/components/HomeComponents/Tasks/Tasks.tsx]: added unsyncedTaskUuids state using sets, adds task UUID to set whenever any field is modified locally, clears all unsynced indicators after successful sync, passes isUnsynced prop to TaskDialog to display visual indicator and total count of unsync tasks on sync button

**Review by ShivaGupta-14 — COMMENTED** [review]
self review completed! open to any suggestions

**Comment by ShivaGupta-14** [comment]
Hi @its-me-abhishek! this is ready for review. please review whenever you get time.
